File: watermelon.py
import cv2
import numpy as np
from scipy.spatial import distance as dist
from imutils import perspective
import imutils

# ================================
# ฟังก์ชัน midpoint (เหมือนของเดิม)
# ================================
def midpoint(ptA, ptB):
    return ((ptA[0] + ptB[0]) * 0.5, (ptA[1] + ptB[1]) * 0.5)

# ================================
# เปิดกล้อง
# ================================
cap = cv2.VideoCapture(0)

# ==============
# NOTE: ผมแก้ตรงนี้
# ของเดิมคุณใช้ while(cap.read()) ซึ่งผิด → ทำให้กล้องค้าง
# ==============
while True:
    ret, frame = cap.read()
    if not ret:
        break

    # resize ลงเพื่อให้เบาเครื่อง
    frame = cv2.resize(frame, None, fx=1, fy=1)

    # =============================
    # แปลงภาพเป็น Gray → Blur → Threshold
    # =============================
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    blur = cv2.GaussianBlur(gray, (15, 15), 0)

    # -------------------------
    # NOTE: โค้ดเดิมมี BUG ผิดสะกด THRESH_BINARY_INV
    # ผมแก้จาก THRESH_BINARY_ INV → THRESH_BINARY_INV
    # -------------------------
    thresh = cv2.adaptiveThreshold(
        blur, 255,
        cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
        cv2.THRESH_BINARY_INV,
        11, 2
    )

    # -------------------------
    # NOTE: แก้ชื่อ kernel (ของเดิม kernal)
    # -------------------------
    kernel = np.ones((3, 3), np.uint8)
    closing = cv2.morphologyEx(thresh, cv2.MORPH_CLOSE, kernel, iterations=3)

    result_img = closing.copy()

    # ============================
    # หา Contours
    # ============================
    cnts = cv2.findContours(result_img, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    cnts = cnts[0] if len(cnts) == 2 else cnts[1]

    hitung_objek = 0

    for cnt in cnts:
        area = cv2.contourArea(cnt)

        # กรองวัตถุที่เล็กเกินไป
        if area < 3000:
            continue

        hitung_objek += 1

        # ======================
        # NOTE: คุณใช้ BoxPoints แบบเก่า ผมแก้ให้เป็นเวอร์ชันใหม่
        # ======================
        box = cv2.minAreaRect(cnt)
        box = cv2.boxPoints(box)
        box = np.array(box, dtype="int")
        box = perspective.order_points(box)

        # วาดกรอบ
        orig = frame.copy()
        cv2.drawContours(orig, [box.astype("int")], -1, (0, 255, 0), 2)

        # ----------------------
        # หา midpoint
        # ----------------------
        (tl, tr, br, bl) = box
        (tltrX, tltrY) = midpoint(tl, tr)
        (blbrX, blbrY) = midpoint(bl, br)

        (tlblX, tlblY) = midpoint(tl, bl)
        (trbrX, trbrY) = midpoint(tr, br)

        # วาดจุด
        cv2.circle(orig, (int(tltrX), int(tltrY)), 5, (0, 255, 0), -1)
        cv2.circle(orig, (int(blbrX), int(blbrY)), 5, (0, 255, 0), -1)
        cv2.circle(orig, (int(tlblX), int(tlblY)), 5, (0, 255, 0), -1)
        cv2.circle(orig, (int(trbrX), int(trbrY)), 5, (0, 255, 0), -1)

        # เส้นกากบาท
        cv2.line(orig, (int(tltrX), int(tltrY)), (int(blbrX), int(blbrY)), (255, 0, 255), 2)
        cv2.line(orig, (int(tlblX), int(tlblY)), (int(trbrX), int(trbrY)), (255, 0, 255), 2)

        # ==========================
        # คำนวณระยะพิกเซล
        # ==========================
        lebar_pixel = dist.euclidean((tltrX, tltrY), (blbrX, blbrY))
        panjang_pixel = dist.euclidean((tlblX, tlblY), (trbrX, trbrY))

        # -------------------------
        # NOTE: คุณคำนวณผิดหลายจุด ผมแก้ดังนี้:
        # - 1 pixel = 1/25.5 cm (ต้องปรับตามกล้องจริง)
        # -------------------------
        cm_per_pixel = 1 / 25.5

        real_width = lebar_pixel * cm_per_pixel
        real_height = panjang_pixel * cm_per_pixel

        # ==========================
        # โมเดลคำนวณน้ำหนักแบบทรงรี (ตามสูตรของคุณ)
        # ==========================
        weight = (0.0019 * ((2 * real_width * (real_height ** 2) * 3.14) / 3)) + 0.2228

        # ==========================
        # แสดงข้อมูลบนจอ
        # ==========================
        cv2.putText(orig, f"L: {real_width:.2f} CM", (10, 30),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 0, 255), 2)

        cv2.putText(orig, f"H: {real_height:.2f} CM", (10, 60),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 0, 255), 2)

        cv2.putText(orig, f"Weight: {weight:.2f} kg", (10, 100),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.9, (255, 20, 127), 2)

        cv2.putText(orig, f"Objects: {hitung_objek}", (10, 140),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.9, (30, 144, 255), 2)

        cv2.putText(orig, "*** Keep 60 cm distance ***",
                    (10, 450), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (50, 50, 50), 2)

        cv2.imshow("Camera", orig)

    # ปิดโปรแกรมเมื่อกด ESC
    if cv2.waitKey(1) & 0xFF == 27:
        break

cap.release()
cv2.destroyAllWindows()

File: watermelon.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <cmath>

using namespace cv;
using namespace std;

// ==============================
// ฟังก์ชัน midpoint 
// ==============================
Point2f midpoint(Point2f a, Point2f b) {
    return Point2f((a.x + b.x) * 0.5f, (a.y + b.y) * 0.5f);
}

int main() {
    VideoCapture cap(0);
    if (!cap.isOpened()) {
        cout << "Cannot open camera" << endl;
        return -1;
    }

    while (true) {
        Mat frame;
        cap >> frame;
        if (frame.empty()) break;

        // Resize
        resize(frame, frame, Size(), 1, 1);

        // =============================
        // Gray → Blur → Threshold
        // =============================
        Mat gray, blurImg, thresh;
        cvtColor(frame, gray, COLOR_BGR2GRAY);
        GaussianBlur(gray, blurImg, Size(15, 15), 0);

        adaptiveThreshold(
            blurImg, thresh, 255,
            ADAPTIVE_THRESH_GAUSSIAN_C,
            THRESH_BINARY_INV,
            11, 2
        );

        // Morph Close
        Mat kernel = getStructuringElement(MORPH_RECT, Size(3, 3));
        Mat closing;
        morphologyEx(thresh, closing, MORPH_CLOSE, kernel, Point(-1, -1), 3);

        // ==========================
        // หา Contours
        // ==========================
        vector<vector<Point>> contours;
        findContours(closing, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);

        int hitung_objek = 0;

        for (auto &cnt : contours) {
            double area = contourArea(cnt);

            if (area < 3000) continue;

            hitung_objek++;

            // ==========================
            // คำนวณกล่องล้อม (minAreaRect)
            // ==========================
            RotatedRect rect = minAreaRect(cnt);
            Point2f box[4];
            rect.points(box);

            // วาดกรอบ
            Mat orig = frame.clone();
            for (int i = 0; i < 4; i++)
                line(orig, box[i], box[(i + 1) % 4], Scalar(0, 255, 0), 2);

            // จัดเรียงมุม (manual)
            // tl, tr, br, bl
            vector<Point2f> pts(box, box + 4);
            sort(pts.begin(), pts.end(),
                [](Point2f a, Point2f b) { return a.x < b.x; });

            Point2f left1 = pts[0], left2 = pts[1];
            Point2f right1 = pts[2], right2 = pts[3];

            Point2f tl = (left1.y < left2.y) ? left1 : left2;
            Point2f bl = (left1.y > left2.y) ? left1 : left2;
            Point2f tr = (right1.y < right2.y) ? right1 : right2;
            Point2f br = (right1.y > right2.y) ? right1 : right2;

            // อ้างอิง midpoint
            Point2f midTop = midpoint(tl, tr);
            Point2f midBottom = midpoint(bl, br);
            Point2f midLeft = midpoint(tl, bl);
            Point2f midRight = midpoint(tr, br);

            // วาดจุด midpoint
            circle(orig, midTop, 5, Scalar(0, 255, 0), -1);
            circle(orig, midBottom,5, Scalar(0, 255, 0), -1);
            circle(orig, midLeft, 5, Scalar(0, 255, 0), -1);
            circle(orig, midRight,5, Scalar(0, 255, 0), -1);

            // เส้นกากบาท
            line(orig, midTop, midBottom, Scalar(255, 0, 255), 2);
            line(orig, midLeft, midRight, Scalar(255, 0, 255), 2);

            // ============================
            // คำนวณระยะพิกเซล
            // ============================
            float width_px = norm(midTop - midBottom);
            float height_px = norm(midLeft - midRight);

            float cm_per_pixel = 1.0f / 25.5f;

            float width_cm = width_px * cm_per_pixel;
            float height_cm = height_px * cm_per_pixel;

            float weight = (0.0019f * ((2 * width_cm * height_cm * height_cm * 3.14f) / 3)) + 0.2228f;

            // =============================
            // แสดงผลบนจอ
            // =============================
            putText(orig, "W: " + to_string(width_cm) + " cm",
                    Point(10, 30), FONT_HERSHEY_SIMPLEX, 0.7,
                    Scalar(0, 0, 255), 2);

            putText(orig, "H: " + to_string(height_cm) + " cm",
                    Point(10, 60), FONT_HERSHEY_SIMPLEX, 0.7,
                    Scalar(0, 0, 255), 2);

            putText(orig, "Weight: " + to_string(weight) + " kg",
                    Point(10, 100), FONT_HERSHEY_SIMPLEX, 0.9,
                    Scalar(255, 20, 127), 2);

            putText(orig, "Objects: " + to_string(hitung_objek),
                    Point(10, 140), FONT_HERSHEY_SIMPLEX, 0.9,
                    Scalar(20, 200, 255), 2);

            putText(orig, "*** Keep 60 cm distance ***",
                    Point(10, 450), FONT_HERSHEY_SIMPLEX, 0.7,
                    Scalar(50, 50, 50), 2);

            imshow("Camera", orig);
        }

        if (waitKey(1) == 27) break;  // ESC
    }

    cap.release();
    destroyAllWindows();
    return 0;
}